staking/entryWriter.go

```diff
@@ -0,0 +1,196 @@
+package staking
+
+import (
+	"errors"
+	"fmt"
+	"github.com/pegnet/pegnet/spr"
+	log "github.com/sirupsen/logrus"
+	"sync"
+
+	"github.com/FactomProject/factom"
+	"github.com/cenkalti/backoff"
+	"github.com/pegnet/pegnet/common"
+	"github.com/zpatrick/go-config"
+)
+
+type IEntryWriter interface {
+	PopulateECAddress() error
+	NextBlockWriter() IEntryWriter
+	SetSPR(spr *spr.StakingPriceRecord)
+	CollectAndWrite(blocking bool)
+	ECBalance() (int64, error)
+}
+
+// EntryWriter writes the SPRs to factom once the staking is done
+type EntryWriter struct {
+	// We need an spr template to make the entries
+	sprTemplate *spr.StakingPriceRecord
+
+	ec     *factom.ECAddress
+	config *config.Config
+
+	Next *EntryWriter
+
+	EntryWritingFunction func() error
+
+	sync.Mutex
+	sync.Once
+}
+
+func NewEntryWriter(config *config.Config) *EntryWriter {
+	w := new(EntryWriter)
+	w.config = config
+	w.EntryWritingFunction = w.writeStakingRecord
+	return w
+}
+
+func (w *EntryWriter) ECBalance() (int64, error) {
+	return factom.GetECBalance(w.ec.String())
+}
+
+// PopulateECAddress only needs to be called once
+func (w *EntryWriter) PopulateECAddress() error {
+	// Get the Entry Credit Address that we need to write our SPR records.
+	if ecadrStr, err := w.config.String("Miner.ECAddress"); err != nil {
+		return err
+	} else {
+		ecAdr, err := factom.FetchECAddress(ecadrStr)
+		if err != nil {
+			return err
+		}
+		w.ec = ecAdr
+	}
+	return nil
+}
+
+// NextBlockWriter gets the next block writer to use for the staker.
+func (w *EntryWriter) NextBlockWriter() IEntryWriter {
+	w.Lock()
+	defer w.Unlock()
+	if w.Next == nil {
+		w.Next = NewEntryWriter(w.config)
+		w.Next.ec = w.ec
+	}
+	return w.Next
+}
+
+// SetSPR is here because we need an spr to create the entry.
+func (w *EntryWriter) SetSPR(spr *spr.StakingPriceRecord) {
+	w.Lock()
+	defer w.Unlock()
+	if w.sprTemplate == nil {
+		w.sprTemplate = spr.CloneEntryData()
+	}
+}
+
+// CollectAndWrite will write the block when we collected all the staker data
+//	The blocking is mainly for unit tests.
+func (w *EntryWriter) CollectAndWrite(blocking bool) {
+	w.Do(func() {
+		if blocking {
+			w.collectAndWrite()
+		} else {
+			go w.collectAndWrite()
+		}
+	})
+}
+
+// collectAndWrite is idempotent
+func (w *EntryWriter) collectAndWrite() {
+	err := w.EntryWritingFunction() // Write to blockchain
+	if err != nil {
+		log.WithError(err).Error("Failed to write staking record")
```

Comment:
should there be some kind of retry or identification of the error? Like not enough EC or network connectivity issue.

Reply:
Retry() is not allowed with any kinds of errors with Staking modules.
In PIP13, we try to write the SPR at minute 1, and write them to SPR chain at minute 8.
All is processed based on synchronization within 10 minutes.

If there is any errors such as insufficient EC, that should be terminated until staker fixes his configuration.

Reply:
Yes, now it makes sense. Thanks for thorough explanation.